Add ISOP expression printing to print.hpp

brainkz opened the pull request:
SOP expressions are useful when constructing the GENLIB library for technology mapping. The functions `print_sop_expression` and `to_sop_expression` convert the complete truth table and produce the SOP expression, e.g., (b&c)|(a&c)|(a&b) for majority-3.

msoeken commented on test/print.cpp:
```diff
@@ -168,4 +168,4 @@ TEST_F( PrintTest, print_kmap )
     }
   }
   infile.close();
```
This change should be reverted, as we expect an empty line at the end of every file.

msoeken commented on include/kitty/print.hpp:
```diff
@@ -478,4 +481,93 @@ std::string anf_to_expression( const ternary_truth_table<TT>& anf )
+    auto bits = cube._bits;
+    auto mask = cube._mask;
+
+    bool brackets = (__builtin_popcount(mask) > 1);
```
```suggestion
    bool brackets = __builtin_popcount(mask) > 1;
```

msoeken commented on include/kitty/print.hpp:
```diff
@@ -478,4 +481,93 @@ std::string anf_to_expression( const ternary_truth_table<TT>& anf )
+  \param os Output stream
+ */
+template<typename TT, typename = std::enable_if_t<is_completely_specified_truth_table<TT>::value>>
+void print_sop_expression(TT tt, std::ostream& os = std::cout )
```
```suggestion
void print_sop_expression( TT tt, std::ostream& os = std::cout )
```

msoeken commented on include/kitty/print.hpp:
```diff
@@ -478,4 +481,93 @@ std::string anf_to_expression( const ternary_truth_table<TT>& anf )
+
+  
+  /* write product terms */ 
+  for (auto cube : cubes)
```
```suggestion
  for ( auto cube : cubes )
```

msoeken commented on include/kitty/print.hpp:
```diff
@@ -478,4 +481,93 @@ std::string anf_to_expression( const ternary_truth_table<TT>& anf )
+
+    bool brackets = (__builtin_popcount(mask) > 1);
+
+    if (!first_cube)
```
In the formatting for this library we expect spaces after opening and before closing parentheses.

msoeken commented on include/kitty/print.hpp:
```diff
@@ -46,6 +47,8 @@
 #include "operations.hpp"
 #include "constructors.hpp"
 
```
No empty line before include.

msoeken reviewed (changes requested):
Thanks for the contribution. I left some small comments that should be addressed before merging.

brainkz commented:
Updated my pull request according to the requested changes:

- No brackets around RHS
`bool brackets = __builtin_popcount( mask ) > 1;`

- Run `clang-format` to correct formatting, particularly:
`void print_sop_expression( TT tt, std::ostream& os = std::cout )`
`for ( auto cube : cubes )`
`if ( !first_cube )`

- Removed empty line before include 
```
#include "algorithm.hpp"
#include "karnaugh_map.hpp"
#include "operations.hpp"
#include "constructors.hpp"
#include "isop.hpp"
```
